ffi/src/primitives.rs

```diff
@@ -288,4 +288,4 @@ macro_rules! primitive_impls {
     )+};
 }
 
-primitive_impls! {u8, u16, u32, u64, i8, i16, i32, i64}
+primitive_impls! {u8, u16, u32, u64, i8, i16, i32, i64, i128, u128}
```

Comment:
[this](/https://github.com/rust-lang/rust/issues/54341#issuecomment-428606606) is not ffi safe as it seems. What I propose is to use `FfiTuple2(u64, u64)`

Reply:
The link is broken.

Reply:
I checked `struct U128([u64; 2]);` and it seems like it passed by value (as `i64 i64`) and returned by reference.

```rust
#[repr(transparent)]
pub struct U128Pair([u64; 2]);

#[no_mangle]
pub unsafe extern "C" fn simple(value: U128Pair) -> U128Pair {
    value
}
```

Produced wasm function: 
```wasm
  (func $simple (type 1) (param i32 i64 i64)
    local.get 0
    local.get 2
    i64.store offset=8
    local.get 0
    local.get 1
    i64.store)
```

It is wroth noticing that rustc produce exactly the same code if we just use `u128`: 

```rust
#[no_mangle]
pub unsafe extern "C" fn simple_plain(value: u128) -> u128 {
    value
}
```

```wasm
  (func $simple_plain (type 1) (param i32 i64 i64)
    local.get 0
    local.get 2
    i64.store offset=8
    local.get 0
    local.get 1
    i64.store)
```

Reply:
ok tnx. So here is the thing. Arrays in Rust are `improper_ctypes`, i.e. they have defined layout, but cannot be used in `extern` functions binding to some C code because in C arrays decay to pointers. To interface C bindings they can either be sent by pointer or wrapped into a struct in which case they'll be passed by value.

Having said this, our library always produces rust FFI code from some rust code and I believe that we shouldn't worry about this. However, if the user wishes to generate c bindings with `cbindgen`, they currently [don't support](https://github.com/eqrion/cbindgen/issues/528) this and offer no workaround

Reply:
To paraphrase, this is something that should be worried about when you have a C library and are generating rust bindings to match it's ABI. If you have a Rust library and are generating C bindings to match it's ABI you can match it on the C side by wrapping an array in a struct. However, this is not something that `cbindgen` supports as of yet. I've opened [a discussion](https://github.com/eqrion/cbindgen/issues/777) to see if this is something that can be supported

Also, note that `u128`/`i128` [have](https://doc.rust-lang.org/reference/type-layout.html#primitive-data-layout) defined layouts

Reply:
I might have been wrong here, I didn't think about calling convention.
Let's move the discussion to #2622 and #2629